File: cpp/classic/greedy_algorithm/IntervalCoverage.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class IntervalCoverage {
public:
    int solution(vector<vector<int>> &values, vector<int> &target) {
        sort(values.begin(), values.end(), cmp);

        int start = target[0];
        int end = target[1];
        int index = 0;
        int result = 0;
        while (start < end) {
            int border = start;
            while (index < (int) values.size() && start >= values[index][0]) {
                if (border < values[index][1]) {
                    border = values[index][1];
                }
                ++index;
            }

            if (border == start) {
                return -1;
            }

            ++result;
            start = border;
        }
        return result;
    }

private:
    static bool cmp(const vector<int> &node1, const vector<int> &node2) {
        return node1[0] < node2[0];
    }
};

File: py/classic/greedy_algorithm/IntervalCoverage.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import functools
from typing import List

class IntervalCoverage:
    def solution(self, values: List[List[int]], target: List[int]) -> int:
        values.sort(key = functools.cmp_to_key(self.__cmp))

        start, end, index, result = target[0], target[1], 0, 0
        while start < end:
            border = start
            while index < len(values) and start >= values[index][0]:
                if border < values[index][1]:
                    border = values[index][1]
                index += 1

            if border == start:
                return -1

            result += 1
            start = border
        return result

    def __cmp(self, node1: List[int], node2: List[int]) -> int:
        return node1[0] - node2[0]
